src/server.js

```diff
@@ -19,97 +19,175 @@ const timeToClose = 1000 * 60 * 60; // 1 hour in milliseconds
-        }
-        socket.to(roomId).emit('clearAll');
+        clearAllStuff(socket, roomId);
     });
 
     socket.on('background', (data) => {
-        console.log('background changed');
-        if (rooms.has(roomId)) {
-            rooms.get(roomId).background = data;
+        changeBackground(roomId, data);
+        redrawLines(roomId, data);
+    });
+}
+
+function disconnectUser(roomId) {
+    if (rooms.has(roomId)) {
+        rooms.get(roomId).users.delete(socket.id);
+
+        if (rooms.get(roomId).users.size == 0) {
+            ConsoleLog.closingRoom(roomId);
+
+            rooms.get(roomId).timeout = setTimeout(
+                deleteRoom,
+                timeToClose,
+                roomId
+            );
         }
-        io.to(roomId).emit('canvas', {
-            lines: rooms.get(roomId).lines,
-            background: data,
-        });
+    }
+}
+
+function draw(roomId, data) {
+    if (rooms.has(roomId)) {
+        rooms.get(roomId).lines.push(data);
+    }
+
+    io.to(roomId).emit('draw', data);
+}
+
+function clearAllStuff(socket, roomId) {
+    if (rooms.has(roomId)) {
+        rooms.get(roomId).lines = [];
+    }
+
+    socket.to(roomId).emit('clearAll');
+}
+
+function changeBackground(roomId, data) {
+    ConsoleLog.backgroundChange();
+
+    if (rooms.has(roomId)) {
+        rooms.get(roomId).background = data;
+    }
+}
+
+function redrawLines(roomId, data) {
+    io.to(roomId).emit('canvas', {
+        lines: rooms.get(roomId).lines,
+        background: data,
     });
 }
 
 function deleteRoom(roomId) {
     if (rooms.has(roomId)) {
         rooms.delete(roomId);
-        console.log('close room', roomId);
+
+        ConsoleLog.closeRoom(roomId);
+        ConsoleLog.roomsCount();
+    }
+}
+
+function roomExsists(data) {
+    return rooms.has(data);
+}
+
+function stopTimeOut(roomId) {
+    if (rooms.get(roomId).timeout != null) {
+        clearTimeout(rooms.get(roomId).timeout);
+        rooms.get(roomId).timeout = null;
+    }
+}
+
+function addUserToRoom(socket, roomId) {
+    rooms.get(roomId).users.add(socket.id);
+    socket.join(roomId);
+    ConsoleLog.userJoin(socket, roomId);
+}
+
+function loadCanvasToUser(socket, roomId) {
+    socket.emit('canvas', {
+        lines: rooms.get(roomId).lines,
+        background: rooms.get(roomId).background,
+    });
+    ConsoleLog.sentLines();
+}
+
+function createRoom(socket, roomId) {
+    rooms.set(roomId, {
+        users: new Set([socket.id]),
+        lines: new Array(),
+        background: null,
+        timeout: null,
+    });
+
+    socket.join(roomId);
+    socket.emit('path', '/' + roomId);
+    ConsoleLog.userJoin(socket, roomId);
+}
+
+function roomProcess(socket, data) {
+    let roomId;
+
+    if (roomExsists(data)) {
+        roomId = data;
+        stopTimeOut(roomId);
+        addUserToRoom(socket, roomId);
+        loadCanvasToUser(socket, roomId);
+    } else {
+        roomId = crypto.randomBytes(9).toString('base64');
```

Comment:
It will be better to replace it with `crypto.randomBytes(10).toString('base32');`
Because now it can generate strings with /

Reply:
Ok.